### DISCUSSION TITLE
initiate failed: establishing CHILD_SA 'esp-transport' failed

### DISCUSSION THREAD

**Discussion by KushalGowda-Visteon** [strongswan/strongswan#2559] [Q&A]
I am establishing ESP transport connection between android and QNX. I am getting this error while initiating child in android side(adb shell.)
/vendor/bin # **swanctl --initiate --child esp-transport**
12[CFG] vici initiate CHILD_SA 'esp-transport'
16[IKE] queueing CHILD_CREATE task
16[IKE] activating new tasks
16[IKE]   activating CHILD_CREATE task
16[IKE] establishing CHILD_SA esp-transport{1}
16[KNL]   SADB_EXT_SA
16[KNL]   SADB_EXT_LIFETIME_HARD
16[KNL]   SADB_EXT_LIFETIME_SOFT
16[KNL]   SADB_EXT_LIFETIME_CURRENT
16[KNL]   SADB_EXT_ADDRESS_SRC
16[KNL]   SADB_EXT_ADDRESS_DST
16[KNL]   SADB_X_EXT_SA2
16[KNL] got SPI c8e937f5
16[ENC] generating CREATE_CHILD_SA request 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
[IKE] establishing CHILD_SA esp-transport{1}
[ENC] generating CREATE_CHILD_SA request 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
16[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (208 bytes)
[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (208 bytes)
11[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (208 bytes)
[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (208 bytes)
11[ENC] parsed CREATE_CHILD_SA response 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
11[IKE] received USE_TRANSPORT_MODE notify
11[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED notify
[ENC] parsed CREATE_CHILD_SA response 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
11[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
11[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
11[KNL] deleting SAD entry with SPI c8e937f5
11[KNL] deleted SAD entry with SPI c8e937f5
11[KNL] adding SAD entry with SPI c8e937f5 and reqid {1}
11[KNL]   using encryption algorithm AES_GCM_16 with key size 160
11[KNL] unable to add SAD entry with SPI c8e937f5: Function not implemented (38)
11[KNL] adding SAD entry with SPI c7e472b8 and reqid {1}
11[KNL]   using encryption algorithm AES_GCM_16 with key size 160
[KNL] unable to add SAD entry with SPI c8e937f5: Function not implemented (38)
11[KNL] unable to add SAD entry with SPI c7e472b8: Function not implemented (38)
[KNL] unable to add SAD entry with SPI c7e472b8: Function not implemented (38)
11[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
11[IKE] failed to establish CHILD_SA, keeping IKE_SA
11[IKE] reinitiating already active tasks
11[IKE]   CHILD_CREATE task
11[IKE] sending DELETE for ESP CHILD_SA with SPI c8e937f5
[IKE] failed to establish CHILD_SA, keeping IKE_SA
[IKE] sending DELETE for ESP CHILD_SA with SPI c8e937f5
11[ENC] generating INFORMATIONAL request 3 [ D ]
11[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (80 bytes)
[ENC] generating INFORMATIONAL request 3 [ D ]
[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (80 bytes)
16[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (80 bytes)
[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (80 bytes)
16[ENC] parsed INFORMATIONAL response 3 [ D ]
16[KNL] deleting policy 100.64.10.3/32[17/22800] === 100.64.10.4/32[17/22800] in
16[KNL] deleting policy 100.64.10.3/32[17/22800] === 100.64.10.4/32[17/22800] in failed, not found
[ENC] parsed INFORMATIONAL response 3 [ D ]
16[KNL] deleting SAD entry with SPI c8e937f5
16[KNL] unable to delete SAD entry with SPI c8e937f5: No such process (3)
16[KNL] deleting SAD entry with SPI c7e472b8
16[KNL] unable to delete SAD entry with SPI c7e472b8: No such process (3)
16[IKE] activating new tasks
16[IKE] nothing to initiate
[KNL] deleting policy 100.64.10.3/32[17/22800] === 100.64.10.4/32[17/22800] in failed, not found
[KNL] unable to delete SAD entry with SPI c8e937f5: No such process (3)
[KNL] unable to delete SAD entry with SPI c7e472b8: No such process
**initiate failed: establishing CHILD_SA 'esp-transport' failed**

Below folder contains swanctl.conf and prior logs to above error(swanctl --load-all) (swanctl --initiate --ike ubuntu-to-b)
[Android-QNX.zip](https://github.com/user-attachments/files/18020323/Android-QNX.zip)
I am using below android architecture:
file /vendor/bin/charon
/vendor/bin/charon: ELF shared object, 64-bit LSB arm64, dynamic (/system/bin/linker64), for Android 31, built by NDK r23b (7779620), not stripped

Any idea?

**Comment by tobiasbrunner** [comment]
> 11[KNL] adding SAD entry with SPI c8e937f5 and reqid {1}
> 11[KNL] using encryption algorithm AES_GCM_16 with key size 160
> 11[KNL] unable to add SAD entry with SPI c8e937f5: Function not implemented (38)

Your kernel might not be supporting AES-GCM (or there is some other [kernel module](https://docs.strongswan.org/docs/latest/install/kernelModules.html) missing).

**Reply by KushalGowda-Visteon** [reply]
This is list of algs:
exynosauto9_discovery_jlrcccm:/vendor/bin # swanctl --list-algs

aead:
  **AES_GCM_16[openssl]**
  AES_GCM_12[openssl]
  AES_GCM_8[openssl]
  AES_CCM_16[openssl]
  AES_CCM_12[openssl]
  AES_CCM_8[openssl]
  CHACHA20_POLY1305[openssl]
  
  We can see **AES_GCM_16[openssl]** supported right? Also any idea which missing modules to include?

**Reply by tobiasbrunner** [reply]
> This is list of algs:
> exynosauto9_discovery_jlrcccm:/vendor/bin # swanctl --list-algs

These algorithms are the ones loaded by the IKE daemon, so they are only relevant for IKE. ESP traffic is handled by the kernel, so it's the modules that are loaded (or loadable) there that are relevant (`/proc/crypto` can give you some indications).

**Reply by KushalGowda-Visteon** [reply]
Okay got it. Now here in /proc/crypto
/proc # cat crypto | grep "aes"
name         : rfc3686(ctr(aes))
driver       : rfc3686(ctr-aes-ce)
name         : cbcmac(aes)
driver       : cbcmac-aes-ce
name         : xcbc(aes)
driver       : xcbc-aes-ce
name         : cmac(aes)
driver       : cmac-aes-ce
name         : essiv(cbc(aes),sha256)
driver       : essiv-cbc-aes-sha256-ce
name         : cts(cbc(aes))
driver       : cts-cbc-aes-ce
name         : xts(aes)
driver       : xts-aes-ce
name         : xctr(aes)
driver       : xctr-aes-ce
name         : ctr(aes)
driver       : ctr-aes-ce
name         : cbc(aes)
driver       : cbc-aes-ce
name         : ecb(aes)
driver       : ecb-aes-ce
name         : aes
driver       : aes-ce
name         : aes
driver       : aes-generic
This above list doesnt include aes_gcm.
also cat crypto | grep "gcm"
returns empty.

So we are enabling
CONFIG_CRYPTO_AES=y
CONFIG_CRYPTO_GCM=y

Hope we are going in right direction?

**Reply by tobiasbrunner** [reply]
Yes, sounds reasonable.

**Reply by KushalGowda-Visteon** [reply]
Even though required kernel modules are enabled(checked through this path ( /proc/config.gz ) 
CONFIG_NET_KEY=y
CONFIG_XFRM=y
CONFIG_XFRM_USER=y
CONFIG_XFRM_ALGO=y
CONFIG_NETFILTER_XT_TARGET_TPROXY=y
CONFIG_NETFILTER_XT_MARK=y
CONFIG_IP_NF_IPTABLES=y
CONFIG_NETFILTER_XT_TARGET_LOG=y
CONFIG_INET_XFRM_MODE_TRANSPORT=y
CONFIG_INET_XFRM_MODE_TUNNEL=y
CONFIG_NET=y
CONFIG_NETLINK=y
CONFIG_NETFILTER_NETLINK=y
CONFIG_XFRM_IPCOMP=y
CONFIG_INET_XFRM_MODE_UDP=y
CONFIG_CRYPTO_AES=y
CONFIG_CRYPTO_SHA256=y
CONFIG_CRYPTO_HMAC=y
CONFIG_CRYPTO_USER=y
CONFIG_INET=y
CONFIG_IPV6=y
CONFIG_INET_AH=y
CONFIG_INET6_AH=y
CONFIG_INET_ESP=y
CONFIG_INET6_ESP=y
CONFIG_INET_XFRM_MODE_TRANSPORT=y
CONFIG_INET6_XFRM_MODE_TRANSPORT=y
CONFIG_CRYPTO_AES=y
CONFIG_CRYPTO_CBC=y
CONFIG_CRYPTO_ECB=y
CONFIG_CRYPTO_GCM=y
CONFIG_CRYPTO_HMAC=y
CONFIG_CRYPTO_MD5=y
CONFIG_CRYPTO_NULL=y
CONFIG_CRYPTO_SHA1=y
CONFIG_CRYPTO_SHA256=y
CONFIG_CRYPTO_SHA512=y
CONFIG_INET_IPCOMP=y
CONFIG_INET_XFRM_MODE_TRANSPORT=y
CONFIG_INET_XFRM_MODE_TUNNEL=y
CONFIG_INET_XFRM_MODE_BEET=y
CONFIG_INET6_XFRM_MODE_TRANSPORT=y
CONFIG_INET6_XFRM_MODE_TUNNEL=y
CONFIG_INET6_XFRM_MODE_BEET=y
CONFIG_IP_ADVANCED_ROUTER=y
CONFIG_IP_MULTIPLE_TABLES=y
CONFIG_INET6_IPCOMP=y
CONFIG_IPV6_MULTIPLE_TABLES=y
CONFIG_NETFILTER=y
CONFIG_NETFILTER_XTABLES=y
CONFIG_NETFILTER_XT_MATCH_POLICY=y

loaded plugins:
./configure --disable-shared --disable-systemd --enable-static --enable-monolithic --enable-gmp --enable-addrblock --enable-kernel-pfkey --enable-whitelist --enable-vici \
--enable-socket-default --enable-openssl --enable-socket-dynamic --enable-eap-mschapv2 --enable-ikev1 --enable-ikev2 --enable-led --enable-android-log

still facing the issue:
exynosauto9_discovery_jlrcccm:/vendor/bin # ./swanctl --initiate --child esp-transport
[IKE] establishing CHILD_SA esp-transport{1}
[ENC] generating CREATE_CHILD_SA request 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (208 bytes)
[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (208 bytes)
[ENC] parsed CREATE_CHILD_SA response 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
[KNL] unable to add SAD entry with SPI c3b5e0cd: Function not implemented (38)
[KNL] unable to add SAD entry with SPI c3fc43b4: Function not implemented (38)
[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
[IKE] failed to establish CHILD_SA, keeping IKE_SA
[IKE] sending DELETE for ESP CHILD_SA with SPI c3b5e0cd
[ENC] generating INFORMATIONAL request 3 [ D ]
[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (80 bytes)
[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (80 bytes)
[ENC] parsed INFORMATIONAL response 3 [ D ]
[KNL] deleting policy 100.64.10.3/32[17/22800] === 100.64.10.4/32[17/22800] in failed, not found
initiate failed: establishing CHILD_SA 'esp-transport' failed

**Reply by tobiasbrunner** [reply]
Apparently something is still missing in the kernel. Maybe IV generators like `CONFIG_CRYPTO_SEQIV` (they should be enabled by `CONFIG_XFRM_ESP`, though)? What kernel version do you use?

**Reply by KushalGowda-Visteon** [reply]
Kernel version:
uname -r
5.15.74-android13-8-g9dcd2fcb79ed-dirty

Also after removing **gcm16-esn** from
 esp_proposals = aes128gcm16-esn
 new swanctl.conf:
 `
    connections {
    android-qnx {
        local_addrs = 100.64.10.4
        remote_addrs = 100.64.10.3
        version = 2
        proposals = aes128-sha256-prfsha256-ecp256

        local {
            auth = psk
            id = 100.64.10.4
        }
        remote {
            auth = psk
            id = 100.64.10.3
        }

        children {
            esp-transport {
                mode = transport
                esp_proposals = aes128
                local_ts = 100.64.10.4[17/any]
                remote_ts =  100.64.10.3[17/22000]
            }
        }
    }
}

secrets {
    ike {
        id = 100.64.10.3
        secret = 0x01010101010101010101010101010101
    }
}
`
 I am getting success logs(not sure whether its completely successful connection though)

exynosauto9_discovery_jlrcccm:/vendor/bin # swanctl --initiate --child esp-transport
04[CFG] vici initiate CHILD_SA 'esp-transport'
12[IKE] queueing CHILD_CREATE task
12[IKE] activating new tasks
12[IKE]   activating CHILD_CREATE task
12[IKE] establishing CHILD_SA esp-transport{1}
12[KNL]   SADB_EXT_SA
12[KNL]   SADB_EXT_LIFETIME_HARD
12[KNL]   SADB_EXT_LIFETIME_SOFT
12[KNL]   SADB_EXT_LIFETIME_CURRENT
12[KNL]   SADB_EXT_ADDRESS_SRC
12[KNL]   SADB_EXT_ADDRESS_DST
12[KNL]   SADB_X_EXT_SA2
12[KNL] got SPI c4e6e4fd
12[ENC] generating CREATE_CHILD_SA request 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
[IKE] establishing CHILD_SA esp-transport{1}
[ENC] generating CREATE_CHILD_SA request 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
12[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (208 bytes)
[NET] sending packet: from 100.64.10.4[4500] to 100.64.10.3[4500] (208 bytes)
13[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (208 bytes)
[NET] received packet: from 100.64.10.3[4500] to 100.64.10.4[4500] (208 bytes)
13[ENC] parsed CREATE_CHILD_SA response 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
[ENC] parsed CREATE_CHILD_SA response 2 [ N(USE_TRANSP) N(ESP_TFC_PAD_N) SA No TSi TSr ]
13[IKE] received USE_TRANSPORT_MODE notify
13[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED notify
13[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
13[CFG] selected proposal: ESP:AES_CBC_128/NO_EXT_SEQ
[CFG] selected proposal: ESP:AES_CBC_128/NO_EXT_SEQ
13[KNL] deleting SAD entry with SPI c4e6e4fd
13[KNL] deleted SAD entry with SPI c4e6e4fd
13[KNL] adding SAD entry with SPI c4e6e4fd and reqid {1}
13[KNL]   using encryption algorithm AES_CBC with key size 128
13[KNL] adding SAD entry with SPI c0ad4c3c and reqid {1}
13[KNL]   using encryption algorithm AES_CBC with key size 128
13[KNL] adding policy 100.64.10.3/32[17/22800] === 100.64.10.4/32[17/22800] in
13[KNL] policy already exists, try to update it
13[KNL]   SADB_EXT_ADDRESS_SRC
13[KNL]   SADB_EXT_ADDRESS_DST
13[KNL] [KNL] policy already exists, try to update it
  SADB_EXT_LIFETIME_HARD
13[KNL]   SADB_EXT_LIFETIME_SOFT
13[KNL]   SADB_EXT_LIFETIME_CURRENT
13[KNL]   SADB_X_EXT_POLICY
13[KNL] adding policy 100.64.10.4/32[17/22800] === 100.64.10.3/32[17/22800] out
13[KNL] policy already exists, try to update it
13[KNL]   SADB_EXT_ADDRESS_SRC
13[KNL]   SADB_EXT_ADDRESS_DST
13[KNL]   SADB_EXT_LIFETIME_HARD
13[KNL]   SADB_EXT_LIFETIME_SOFT
13[KNL]   SADB_EXT_LIFETIME_CURRENT
13[KNL]   SADB_X_EXT_POLICY
[KNL] policy already exists, try to update it
13[IKE] CHILD_SA esp-transport{1} established with SPIs c4e6e4fd_i c0ad4c3c_o and TS 100.64.10.4/32[17/22800] === 100.64.10.3/32[17/22800]
[IKE] CHILD_SA esp-transport{1} established with SPIs c4e6e4fd_i c0ad4c3c_o and TS 100.64.10.4/32[17/22800] === 100.64.10.3/32[17/22800]
**initiate completed successfully**
13[IKE] activating new tasks
13[IKE] nothing to initiate

Does this mean without gcm, Ipsec conn is established?

**Reply by tobiasbrunner** [reply]
> Does this mean without gcm, Ipsec conn is established?

Yes, looks like it. So it seems that there is an issue with AES-GCM in your kernel. As mentioned above, the IV generator could be an issue as CBC uses a different one than GCM. On the other hand, `CONFIG_XFRM_ESP` should enable that, which in turn should be enabled by `CONFIG_INET_ESP`. So not sure what exactly is going wrong with your kernel.
